bitcash/crypto.py

```diff
@@ -20,3 +23,110 @@ def ripemd160_sha256(bytestr):
 
 
 hash160 = ripemd160_sha256
+
+
+def sha512(bytestr):
+    return _sha512(bytestr).digest()
+
+
+# ECIES encryption/decryption methods; AES-128-CBC with PKCS7 is used
+# as the cipher; hmac-sha256 is used as the mac
+# Implementation follows the Electron-Cash implementation of the same
+def aes_encrypt_with_iv(key, iv, data):
+    """Provides AES-CBC encryption of data with key and iv
+
+    :param key: key for the encryption
+    :type key: ``bytes``
+    :param iv: Initialisation vector for the encryption
+    :type iv: ``bytes``
+    :param data: the data to be encrypted
+    :type data: ``bytes``
+    """
+    aes_cbc = pyaes.AESModeOfOperationCBC(key, iv=iv)
+    aes = pyaes.Encrypter(aes_cbc)
+    # empty aes.feed() flushes buffer
+    return aes.feed(data) + aes.feed()
+
+
+def aes_decrypt_with_iv(key, iv, data):
+    """Provides AES-CBC decryption of data with key and iv
+
+    :param key: key for the decryption
+    :type key: ``bytes``
+    :param iv: Initialisation vector for the decryption
+    :type iv: ``bytes``
+    :param data: the data to be decrypted
+    :type data: ``bytes``
+    :raises ValueError: if incorrect ``key`` or ``iv`` give a padding error
+                        during decryption
+    """
+    # assert_bytes(key, iv, data)
+    aes_cbc = pyaes.AESModeOfOperationCBC(key, iv=iv)
+    aes = pyaes.Decrypter(aes_cbc)
+    try:
+        # empty aes.feed() flushes buffer
+        return aes.feed(data) + aes.feed()
+    except ValueError:
+        raise ValueError('Invalid key or iv')
+
+
+def ecies_encrypt(message, pubkey):
+    """Encrypt message with the given pubkey
+
+    :param message: the message to be encrypted
+    :type message: ``bytes``
+    :param pubkey: the public key to be used
+    :type pubkey: ``bytes``
+    """
+    pk = ECPublicKey(pubkey)
+
+    # random key
+    ephemeral = ECPrivateKey()
+    ecdh_key = pk.multiply(ephemeral.secret).format()
+    key = sha512(ecdh_key)
+
+    # aes key and iv, and hmac key
+    iv, key_e, key_m = key[0:16], key[16:32], key[32:]
+    ciphertext = aes_encrypt_with_iv(key_e, iv, message)
+    encrypted = (
+        b'BIE1'
+        + ephemeral.public_key.format()
+        + ciphertext
+    )
+    mac = hmac.new(key_m, encrypted, _sha256).digest()
+
+    return base64.b64encode(encrypted + mac)
+
+
+def ecies_decrypt(encrypted, secret):
```

Comment:
Might consider making all functions except `ecies_*` private, or do `bitcash/_crypto.py` and only expose the wallet methods. Either way is good with me.

Reply:
Made aes_* functions private, since they're only used by ecies_* functions  -- which should be public. But the other functions should be left public, since they expose hashlib functions in a simpler way, as used by Bitcash in other functions/classes.